server: Set external address based on completed serving config

astefanutti opened the pull request:
## Summary

This PR makes sure the external address admission initialiser relies on a completed configuration, with the default secure serving port.

The logic is taken from the generic options completion method. While it fixes the issues below, there may be a better, more idiomatic, way to fully complete the options.

## Related issue(s)

Fixes #853 
Fixes #854

ncdc commented:
@astefanutti thanks for finding this! What do you think about this alternative approach (note, was a quick hack, so names may need some tidying up):

```diff
diff --git a/pkg/admission/clusterworkspaceshard/admission.go b/pkg/admission/clusterworkspaceshard/admission.go
index bcc35b38..52a5cb0a 100644
--- a/pkg/admission/clusterworkspaceshard/admission.go
+++ b/pkg/admission/clusterworkspaceshard/admission.go
@@ -51,7 +51,7 @@ func Register(plugins *admission.Plugins) {
 type clusterWorkspaceShard struct {
 	*admission.Handler
 
-	externalAddress string
+	externalAddress func() string
 }
 
 // Ensure that the required admission interfaces are implemented.
@@ -102,7 +102,7 @@ func (o *clusterWorkspaceShard) Admit(ctx context.Context, a admission.Attribute
 	}
 
 	if cws.Spec.BaseURL == "" {
-		cws.Spec.BaseURL = "https://" + o.externalAddress
+		cws.Spec.BaseURL = "https://" + o.externalAddress()
 	}
 	if cws.Spec.ExternalURL == "" {
 		cws.Spec.ExternalURL = cws.Spec.BaseURL
@@ -117,6 +117,6 @@ func (o *clusterWorkspaceShard) Admit(ctx context.Context, a admission.Attribute
 	return nil
 }
 
-func (o *clusterWorkspaceShard) SetExternalAddress(externalAddress string) {
+func (o *clusterWorkspaceShard) SetExternalAddress(externalAddress func() string) {
 	o.externalAddress = externalAddress
 }
diff --git a/pkg/admission/initializers/initializer.go b/pkg/admission/initializers/initializer.go
index 3b80a180..8dc3b26f 100644
--- a/pkg/admission/initializers/initializer.go
+++ b/pkg/admission/initializers/initializer.go
@@ -87,19 +87,19 @@ func (i *kcpClusterClientInitializer) Initialize(plugin admission.Interface) {
 // NewExternalAddressInitializer returns an admission plugin initializer that injects
 // the external address of the shard into the admission plugin.
 func NewExternalAddressInitializer(
-	externalAddress string,
+	externalAddressProvider func() string,
 ) *externalAddressInitializer {
 	return &externalAddressInitializer{
-		externalAddress: externalAddress,
+		externalAddressProvider: externalAddressProvider,
 	}
 }
 
 type externalAddressInitializer struct {
-	externalAddress string
+	externalAddressProvider func() string
 }
 
 func (i *externalAddressInitializer) Initialize(plugin admission.Interface) {
 	if wants, ok := plugin.(WantExternalAddress); ok {
-		wants.SetExternalAddress(i.externalAddress)
+		wants.SetExternalAddress(i.externalAddressProvider)
 	}
 }
diff --git a/pkg/admission/initializers/interfaces.go b/pkg/admission/initializers/interfaces.go
index ecc94128..c5f04bb8 100644
--- a/pkg/admission/initializers/interfaces.go
+++ b/pkg/admission/initializers/interfaces.go
@@ -44,5 +44,5 @@ type WantsKcpClusterClient interface {
 // WantExternalAddress interface should be implemented by admission plugins
 // that want to have a external shard address injected.
 type WantExternalAddress interface {
-	SetExternalAddress(externalAddress string)
+	SetExternalAddress(externalAddress func() string)
 }
diff --git a/pkg/server/server.go b/pkg/server/server.go
index 0f8dff21..62ab206c 100644
--- a/pkg/server/server.go
+++ b/pkg/server/server.go
@@ -242,7 +242,7 @@ func (s *Server) Run(ctx context.Context) error {
 		kcpadmissioninitializers.NewKcpInformersInitializer(s.kcpSharedInformerFactory),
 		kcpadmissioninitializers.NewKubeClusterClientInitializer(kubeClusterClient),
 		kcpadmissioninitializers.NewKcpClusterClientInitializer(kcpClusterClient),
-		kcpadmissioninitializers.NewExternalAddressInitializer(genericConfig.ExternalAddress),
+		kcpadmissioninitializers.NewExternalAddressInitializer(func() string { return genericConfig.ExternalAddress }),
 	}
 
 	apisConfig, err := genericcontrolplane.CreateKubeAPIServerConfig(genericConfig, s.options.GenericControlPlane, s.kubeSharedInformerFactory, admissionPluginInitializers, storageFactory)
```

ncdc commented:
This lets the normal `Complete()` flow set the external address like it normally does (to get the port in there if needed), and makes it lazily evaluated for the clusterworkspaceshard

astefanutti commented:
@ncdc ah that lazy initializer is exactly what I tried in the first place, but I didn't get the `clusterWorkspaceShard` part right for some reasons :) Let me rework it now that you've showed me the complete solution 😄 Thanks a lot!

sttts commented on pkg/server/server.go:
```diff
@@ -238,6 +240,19 @@ func (s *Server) Run(ctx context.Context) error {
+		}
+		_, port, err := genericConfig.SecureServing.HostPort()
+		if err != nil {
+			klog.Fatalf("cannot derive external address from the secure port: %v", err)
```
return error, don't klog.Fatal.

sttts commented on pkg/server/server.go:
```diff
@@ -238,6 +240,19 @@ func (s *Server) Run(ctx context.Context) error {
+		if err != nil {
+			klog.Fatalf("cannot derive external address from the secure port: %v", err)
+		}
+		genericConfig.ExternalAddress = net.JoinHostPort(genericConfig.ExternalAddress, strconv.Itoa(port))
```
is this a bug in kube?

astefanutti replied:
I don't think so. It's the admission plugin that's initialised in KCP before the `Complete` method is actually called on the generic options.

astefanutti commented:
@ncdc @sttts I've updated the PR. I went for using `*string`, instead of a _provider_/_supplier_ function approach, that simplifies things even more, and added a couple of unit and e2e tests.

ncdc commented on pkg/server/server.go:
```diff
@@ -242,7 +242,9 @@ func (s *Server) Run(ctx context.Context) error {
-		kcpadmissioninitializers.NewExternalAddressInitializer(genericConfig.ExternalAddress),
+		// The external address pointer is passed as its value may be updated
+		// with the default secure port, when the config is completed.
+		kcpadmissioninitializers.NewExternalAddressInitializer(&genericConfig.ExternalAddress),
```
Passing in the pointer means that any admission plugin could potentially mutate genericConfig.ExternalAddress. We presumably wouldn't code that, but it is possible. Do you think we should guard against that?

ncdc commented on pkg/admission/clusterworkspaceshard/admission_test.go:
```diff
@@ -309,8 +309,82 @@ func TestAdmit(t *testing.T) {
+	}
+}
+
+func TestAdmitWithoutExternalAddress(t *testing.T) {
```
Do you think you could combine this into the previous test? Maybe add a bool to the test struct `provideExternalAddress` and have that set/not set `externalAddressProvider` in `o`?

astefanutti replied:
Good idea. I've added the plugin under test to the struct, so the test cases can still be run in parallel if we ever want it to.

astefanutti commented on pkg/admission/clusterworkspaceshard/admission_test.go:
```diff
@@ -69,14 +69,28 @@ func updateAttr(ws, old *tenancyv1alpha1.ClusterWorkspaceShard) admission.Attrib
+
 	tests := []struct {
 		name        string
+		plugin      *clusterWorkspaceShard
```
If I understanding it correctly, wouldn't this approach prevent from running the test cases in parallel in the future? Is that OK?

ncdc replied:
I don't think so. Each subtest is still constructing its own `clusterWorkspaceShard` instance. Right?

astefanutti replied:
Ah you're right, I've overlooked it, sorry :( This should now be closer to your expectation.
